lib/puppet/transaction.rb

```diff
@@ -338,17 +353,30 @@ def resources_by_provider(type_name, provider_name)
   # types, just providers.
   def prefetch(provider_class, resources)
     type_name = provider_class.resource_type.name
-    return if @prefetched_providers[type_name][provider_class.name]
+    return if @prefetched_providers[type_name][provider_class.name] or
+      @prefetch_failed_providers[type_name][provider_class.name]
     Puppet.debug "Prefetching #{provider_class.name} resources for #{type_name}"
     begin
       provider_class.prefetch(resources)
-    rescue LoadError, Puppet::MissingCommand => detail
+    rescue Exception => detail
+      message = _("Could not prefetch %{type_name} provider '%{name}': %{detail}") % { type_name: type_name, name: provider_class.name, detail: detail }
+      if(!detail.is_a?(LoadError) && !detail.is_a?(Puppet::MissingCommand))
```

Comment:
Given that we're notting everything would it read better as

unless detail.is_a?(LoadError) or detail.is_a?(Puppet::MissingCommand)

Reply:
I'd disagree with changing this to an `unless`. The rule-of-thumb that I've followed is that if more than one condition is involved, always use `if`. Keeping track of the inversions around the conditions themselves, and the interaction with `and`/`or`/`||`/`&&` quickly grows out of hand.

I'd definitely remove the outer parens, though: `if !detail.is_a?(LoadError) && !detail.is_a?(Puppet::MissingCommand)`